relay-server/src/middlewares/body_timing.rs

```diff
@@ -0,0 +1,288 @@
+use crate::statsd::RelayTimers;
+use axum::body::{Body, HttpBody};
+use axum::http::Request;
+use hyper::body::Frame;
+use relay_statsd::metric;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::time::Instant;
+use tower::{Layer, Service};
+
+/// Middleware layer that wraps the request [`Body`] to measure body reading duration.
+///
+/// Metrics are tagged with a completion status to track whether the body:
+/// - was read to completion ("completed")
+/// - failed during reading ("failed")
+/// - was dropped before completion ("dropped")
+///
+/// The reported size reflects the actual number of bytes read, which may differ from
+/// the `Content-Length` header. For bodies that fail or are dropped mid-reading,
+/// the size represents only the bytes successfully read before the interruption.
+#[derive(Clone, Debug)]
+pub struct BodyTimingLayer;
+
+impl BodyTimingLayer {
+    pub fn new() -> Self {
+        Self {}
+    }
+}
```

Comment:
Surprised clippy doesn't complain here, I'd just derive default and not have a constructor, especially since this struct can be created by anyone already (no body).